test/others/criu-ns/run.py

```diff
@@ -0,0 +1,181 @@
+#!/usr/bin/env python
+
+import os
+import pty
+import shutil
+import subprocess
+import sys
+import time
+
+
+CRIU_BIN = "../../../criu/criu"
+CRIU_NS = "../../../scripts/criu-ns"
+IMG_DIR = "dumpdir"
+PIDFILE = "pidfile"
+
+
+def check_dumpdir():
+    if os.path.isdir(IMG_DIR):
+        shutil.rmtree(IMG_DIR)
+    os.mkdir(IMG_DIR, 0o755)
+
+
+def create_pty():
+    fd_m, fd_s = pty.openpty()
+    return (os.fdopen(fd_m, "wb"), os.fdopen(fd_s, "wb"))
+
+
+def create_isolated_dumpee():
+    with open("running", "w") as file:
+        file.name
+    fd_m, fd_s = create_pty()
+    pid = os.fork()
+    fd_m.close()
+    if pid == 0:
+        os.setsid()
+        os.dup2(fd_s.fileno(), 0)
+        os.dup2(fd_s.fileno(), 1)
+        os.dup2(fd_s.fileno(), 2)
+        while True:
+            if not os.access("running", os.F_OK):
+                sys.exit(0)
+            time.sleep(1)
+    return (pid, fd_s)
+
+
+def criu_ns_dump(pid, shell_job=False):
+    if shell_job:
+        cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v4",
+               "--shell-job", "-t", str(pid), "--criu-binary",
+               CRIU_BIN]
+    else:
+        cmd = [CRIU_NS, "dump", "-D", "dumpdir", "-v4",
+               "-t", str(pid), "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    return ret
+
+
+def criu_ns_restore(shell_job=False, restore_detached=False):
+    if shell_job:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--shell-job", "--criu-binary", CRIU_BIN]
+    else:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--criu-binary", CRIU_BIN]
+    if restore_detached:
+        cmd = [CRIU_NS, "restore", "-D", "dumpdir", "-v4",
+               "--restore-detached", "--pidfile", PIDFILE,
+               "--criu-binary", CRIU_BIN]
+    ret = subprocess.Popen(cmd).wait()
+    return ret
+
+
+def test_dump_and_restore_with_shell_job():
+    """Test criu-ns dump and restore with --shell-job option"""
+    check_dumpdir()
+    os.setsid()
```

Comment:
Please move this setsid before calling test_dump_and_restore_with_shell_job, this setsid makes process running all testcases session leader, it is counterintuitive to put it into one testcase.

Reply:
done.